iox-#1394 Fix axivion warnings in duration class

ithier opened the pull request:
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] ~~Changelog updated [in the unreleased section][changelog] including API breaking changes~~
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer
<!-- Items in addition to the checklist below that the reviewer should look for -->

1. Fixes warnings for `duration.hpp` and `duration.inl`
3. Adds `operator+=` and `operator-=` operators as required by AUTOSAR guidelines + supporting test cases

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #1394

ithier commented:
@mossmaurice @FerdinandSpitzschnueffler Would you bind reviewing this ?

Build failures do not seem to be related:
1. 503 error on AUTOSAR website causing broken links
2. Something wrong in the Windows build (no idea)

mossmaurice commented:
@ithier Could you please change the commit messages to the format `iox-#1394 Foobar`. Thereby we make sure that the commit is linked with the issue for traceability.

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -60,11 +61,13 @@ inline constexpr Duration Duration::zero() noexcept
 {
     static_assert(std::numeric_limits<T>::is_integer, "only integer types are supported");
 
+    // AXIVION Next Construct AutosarC++19_03-A1.4.3 : Value is an arbitrary integer type, not necessarily unsigned
```
```suggestion
    // AXIVION Next Construct AutosarC++19_03-A1.4.3 : Value is an arbitrary integer type, not necessarily unsigned in a template context
```

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -84,26 +87,29 @@ inline constexpr Duration Duration::fromNanoseconds(const T value) noexcept
+    const auto clampedValue = positiveValueOrClampToZero(value);
     constexpr Duration::Seconds_t MAX_SECONDS_BEFORE_OVERFLOW{std::numeric_limits<Duration::Seconds_t>::max()};
+
+    // AXIVION Next Construct AutosarC++19_03-A0.1.1 : Not dead code as it is platform dependent
```
I don't see Axivion reporting this rule `A project shall not contain instances of non-volatile variables being
given values that are not subsequently used.` ~~Did you mean to suppress `M0.1.2`?~~ You did that below. I suppose this line can be deleted.

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -252,77 +265,68 @@ inline constexpr struct timeval Duration::timeval() const noexcept
     {
         return {std::numeric_limits<SEC_TYPE>::max(), MICROSECS_PER_SEC - 1U};
     }
+    // AXIVION Next Construct AutosarC++19_03-A5.0.9 : Protected by static assertion
```
There is no rule `A5.0.9`. Remove?

ithier replied:
Hmm, where did that come from...

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -252,77 +265,68 @@ inline constexpr struct timeval Duration::timeval() const noexcept
         return {std::numeric_limits<SEC_TYPE>::max(), MICROSECS_PER_SEC - 1U};
     }
+    // AXIVION Next Construct AutosarC++19_03-A5.0.9 : Protected by static assertion
+    // AXIVION Next Construct AutosarC++19_03-M5.0.8 : Protected by static assertion
+    // AXIVION Next Construct AutosarC++19_03-M5.0.9 : Protected by static assertion
```
```suggestion
    // AXIVION Next Construct AutosarC++19_03-M5.0.8, AutosarC++19_03-M5.0.9 : Protected by static assertion
```

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -252,77 +265,68 @@ inline constexpr struct timeval Duration::timeval() const noexcept
     }
-    auto seconds = m_seconds - rhs.m_seconds;
+    Seconds_t seconds{m_seconds - rhs.m_seconds};
+    // AXIVION Next Construct AutosarC++19_03-M0.1.9 : Variable IS used
+    // AXIVION Next Construct AutosarC++19_03-A0.1.1 : Variable IS used
```
```suggestion
    // AXIVION Next Construct AutosarC++19_03-M0.1.9, 03-A0.1.1 : False positive, variable IS used
```

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.hpp:
```diff
@@ -348,10 +366,68 @@ class Duration
+/// @return true if duration equal to rhs
+// AXIVION Next Line AutosarC++19_03-A8.4.7 : Each argument is larger than two words
+constexpr bool operator==(const Duration& lhs, const Duration& rhs) noexcept
+{
+    return (lhs.m_seconds == rhs.m_seconds) && (lhs.m_nanoseconds == rhs.m_nanoseconds);
+}
```
Could you move all the implementations to the `.inl`?

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.hpp:
```diff
@@ -348,10 +366,68 @@ class Duration
+}
+
+/// @brief Greater than or equal to operator
+/// @param[in] rhs is the right hand side of the comparison
```
```suggestion
/// @param[in] lhs is the left hand side of the comparison
/// @param[in] rhs is the right hand side of the comparison
```

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.hpp:
```diff
@@ -348,10 +366,68 @@ class Duration
+}
+
+/// @brief Less than or equal to operator
+/// @param[in] rhs is the right hand side of the comparison
```
```suggestion
/// @param[in] lhs is the left hand side of the comparison
/// @param[in] rhs is the right hand side of the comparison
```

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.hpp:
```diff
@@ -348,10 +366,68 @@ class Duration
+}
+
+/// @brief Greater than operator
+/// @param[in] rhs is the right hand side of the comparison
```
```suggestion
/// @param[in] lhs is the left hand side of the comparison
/// @param[in] rhs is the right hand side of the comparison
```

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -84,26 +87,29 @@ inline constexpr Duration Duration::fromNanoseconds(const T value) noexcept
     constexpr Duration::Seconds_t MAX_SECONDS_BEFORE_OVERFLOW{std::numeric_limits<Duration::Seconds_t>::max()};
+
+    // AXIVION Next Construct AutosarC++19_03-A0.1.1 : Not dead code as it is platform dependent
+    // AXIVION Next Construct AutosarC++19_03-M0.1.2, AutosarC++19_03-M0.1.9, FaultDetection-DeadBranches : False positive. Platform dependent.
```
```suggestion
    // AXIVION Next Construct AutosarC++19_03-M0.1.2, AutosarC++19_03-M0.1.9, FaultDetection-DeadBranches : False positive, whether this is a dead branch is platform dependent.
```

mossmaurice reviewed (commented):
Thanks for creating the PR, LGTM in general! I left a few remarks.

mossmaurice commented:
@ithier 

Re
> Something wrong in the Windows build (no idea)

`units/duration.inl(285,23): error C2248: 'iox::units::Duration::Duration': cannot access protected member declared in class 'iox::units::Duration'`

I think the error is related to this comment:

https://github.com/eclipse-iceoryx/iceoryx/blob/ac127fe6236b4578ba4921b5c1a45782d5f6bb4a/iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.hpp#L320-L321

Can you try calling `createDuration`?

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.hpp:
```diff
@@ -283,17 +287,26 @@ class Duration
 
     // END CONVERSION
 
+    // AXIVION Next Line AutosarC++19_03-A3.9.1 : Use of unsigned long long int in user-defined literals is enforced by the standard
     friend constexpr Duration duration_literals::operator"" _ns(unsigned long long int value) noexcept;
+    // AXIVION Next Line AutosarC++19_03-A3.9.1 : Use of unsigned long long int in user-defined literals is enforced by the standard
     friend constexpr Duration duration_literals::operator"" _us(unsigned long long int value) noexcept;
+    // AXIVION Next Line AutosarC++19_03-A3.9.1 : Use of unsigned long long int in user-defined literals is enforced by the standard
     friend constexpr Duration duration_literals::operator"" _ms(unsigned long long int value) noexcept;
+    // AXIVION Next Line AutosarC++19_03-A3.9.1 : Use of unsigned long long int in user-defined literals is enforced by the standard
     friend constexpr Duration duration_literals::operator"" _s(unsigned long long int value) noexcept;
+    // AXIVION Next Line AutosarC++19_03-A3.9.1 : Use of unsigned long long int in user-defined literals is enforced by the standard
     friend constexpr Duration duration_literals::operator"" _m(unsigned long long int value) noexcept;
+    // AXIVION Next Line AutosarC++19_03-A3.9.1 : Use of unsigned long long int in user-defined literals is enforced by the standard
     friend constexpr Duration duration_literals::operator"" _h(unsigned long long int value) noexcept;
+    // AXIVION Next Line AutosarC++19_03-A3.9.1 : Use of unsigned long long int in user-defined literals is enforced by the standard
     friend constexpr Duration duration_literals::operator"" _d(unsigned long long int value) noexcept;
 
+    // AXIVION Next Construct AutosarC++19_03-A8.4.7 : Argument is larger than two words
```
You could think about using `AXIVION DISABLE /ENABLE STYLE` here as well.

mossmaurice commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -461,53 +466,123 @@ inline constexpr Duration Duration::operator*(const T& rhs) const noexcept
 }
 
+
+/// @brief Equal to operator
+/// @param[in] lhs is the left hand side of the comparison
+/// @param[in] rhs is the right hand side of the comparison
+/// @return true if duration equal to rhs
+// AXIVION Next Line AutosarC++19_03-A8.4.7 : Each argument is larger than two words
```
Please move the doxygen documentation to the `.hpp`.

ithier replied:
@mossmaurice These are not member functions. Are you suggesting to put the prototype for these free functions in the `duration.hpp` header ?

mossmaurice replied:
Correct, that's what I meant. In iceoryx -for templates- declaration and doxygen go into `.hpp` and definition into `.inl`.

mossmaurice reviewed (commented):
Could you squash your last commit 1fe34ec46b17b5105f1944dd0c8cff6f3c49e125. TBH, I'm not a big fan of the "Addressed review findings" commit message, sort of nondescript :sweat_smile:

ithier commented:
@mossmaurice 

> Could you squash your last commit [1fe34ec](https://github.com/eclipse-iceoryx/iceoryx/commit/1fe34ec46b17b5105f1944dd0c8cff6f3c49e125)? TBH, I'm not a big fan of the "Addressed review findings" commit message, sort of nondescript sweat_smile

I use to be the same but I have changed my approach recently because of the following reasons:

1. Addressing each finding and squashing into the appropriate commit requires more effort, sometimes a lot depending on the complexity of the findings
2. The "mess" added to the commit history is only cosmetic and inconsequential
3. The merge commit provides the required granularity for cherry-picking
4. The separate commit makes it easier to re-review an MR and verify that the findings have been addressed

> I'm not a big fan 

Do we have a concrete policy on this approach ? In general I don't think it makes sense to make changes according to personal preference because this will be different with every reviewer.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -60,11 +61,15 @@ inline constexpr Duration Duration::zero() noexcept
 }
 
+// AXIVION Next Construct AutosarC++19_03-A3.9.1 : Use of unsigned long long int in user-defined
+// literals is enforced by the standard
```
This I do not get. Why do we not return a `uint64_t`? Or keep it `T` but clamp to zero (relevant only for signed types of course).

For the latter: we might not want to keep a signed type in this case and it is kind of hard to choose a corresponding unsigned type. It is possible with templates like `signed_type_of<T>` (to be defined), and we can even enforce the non-supported types by making it fail to compile for them.

Disregarding that, why not `uint64_t` if it is non-negative to avoid the comment?

ithier replied:
@MatthiasKillat Ah, good catch, thanks! I had incorrectly treated this case the same as the string literals on lines 524+.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -79,31 +84,35 @@ inline constexpr Duration Duration::fromNanoseconds(const T value) noexcept
-    auto nanoseconds = static_cast<Duration::Nanoseconds_t>((clampedValue % Duration::MICROSECS_PER_SEC)
-                                                            * Duration::NANOSECS_PER_MICROSEC);
-    return Duration{seconds, nanoseconds};
+    const auto clampedValue = positiveValueOrClampToZero(value);
```
Interesting that it does not complain about lack of `{}` initialization here. Better as is anyway, IMO.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -79,31 +84,35 @@ inline constexpr Duration Duration::fromNanoseconds(const T value) noexcept
     constexpr Duration::Seconds_t MAX_SECONDS_BEFORE_OVERFLOW{std::numeric_limits<Duration::Seconds_t>::max()};
+
+    // AXIVION Next Construct AutosarC++19_03-M0.1.2, AutosarC++19_03-M0.1.9,
+    // FaultDetection-DeadBranches : False positive, whether this is a dead branch is platforms
```
Probably because it cannot happen if bit width of `MAX_SECONDS_BEFORE_OVERFLOW` is of at least as large as `uint64_t`. It is difficult to choose an integer type where this is not the case on any platform that does not have larger integers.

I think the comment is fine for this reason, but ideally there should not be the need of a comment at all...

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -145,82 +154,89 @@ inline constexpr Duration Duration::fromDays(const T value) noexcept
     }
 
-    return m_seconds * NANOSECS_PER_SEC + m_nanoseconds;
+    return (m_seconds * NANOSECS_PER_SEC) + m_nanoseconds;
```
**sigh** -.-

Resolve me.

MatthiasKillat commented on iceoryx_hoofs/include/iceoryx_hoofs/internal/units/duration.inl:
```diff
@@ -461,53 +466,99 @@ inline constexpr Duration Duration::operator*(const T& rhs) const noexcept
 }
 
+
+// AXIVION Next Line AutosarC++19_03-A8.4.7 : Each argument is larger than two words
```
This rule apparently does also cause more harm than good or does not work properly ...

Nothing to do. Resolve.

ithier replied:
@MatthiasKillat Yeah, I agree that it doesn't seem to work properly.

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/test_unit_duration.cpp:
```diff
@@ -1647,6 +1647,68 @@ TEST(Duration_test, AddDurationResultsInSaturationFromSeconds)
+
+    sut += otherDuration;
+
+    EXPECT_THAT(sut, Eq(EXPECTED_DURATION));
```
@ithier No need to change, but I prefer `EXPECT_EQ` in those cases. Less verbose and should actually be better performance-wise (the general matcher code invoked here is heavier from inspection of GTest some while ago).

I prefer `EXPECT_THAT` only for complex matching (like contains a substring etc.).

A matter of preference, but this is my rationale.

ithier replied:
@MatthiasKillat Agree, was just following the convention in the file. I will change this one though.

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/test_unit_duration.cpp:
```diff
@@ -1785,6 +1847,59 @@ TEST(Duration_test, SubtractDurationWithSecondsAndNanosecondsCausingReductionOfS
+    EXPECT_THAT(sut, Eq(EXPECTED_DURATION));
+}
+
+TEST(Duration_test, SubtractAssignDurationPastZeroResultsInZero)
```
```suggestion
TEST(Duration_test, SubtractAssignLargerDurationResultsInZero)
```

Only a proposal.

MatthiasKillat commented on iceoryx_hoofs/test/moduletests/test_unit_duration.cpp:
```diff
@@ -1785,6 +1847,59 @@ TEST(Duration_test, SubtractDurationWithSecondsAndNanosecondsCausingReductionOfS
+
+    constexpr Duration EXPECTED_DURATION = createDuration(0U, 0U);
+    auto sut = createDuration(1U, 0U);
+    auto otherDuration = createDuration(2U, 0U);
```
Maybe we also want to check that truncation at 0 happens due to nanoseconds as well (i.e. subtract something larger in the nanosecond range from something smaller). Can even be done in the same test IMO.

ithier replied:
This is tested in the test above: https://github.com/eclipse-iceoryx/iceoryx/pull/1830/files/b45eccf7f12a36dddc0420275ce68c2affff8526#diff-01dd43e4d8b41b14024d0992abde331a2c9b8ec3927488715592516e2e67b8c7R1876

ithier replied:
@MatthiasKillat

MatthiasKillat reviewed (approved):
Looks alright. Optionally one question about a return value and an optional test case could be added.

I think the return value change is probably not in scope here, if it even applies.
